### PR TITLE
Fix disabled no-op module pre-hooks in compile

### LINKED ISSUE #142358: `torch.compiler.disable()` on module hooks will disable `module.compile()`
### 🐛 Describe the bug

`torch.compiler.disable()` on module hooks will disable compile on the whole module when `module.compile()` is used (or via `torch.compile(some_function)(model, some_inputs)`)

```python
import torch
from torch import nn


d_model = 16
model = nn.Sequential(
    nn.Linear(d_model, d_model),
    nn.GELU(),
    nn.Linear(d_model, d_model),
)
model.cuda()

@torch.compiler.disable()
def hook(module, input):
    pass

model.register_forward_pre_hook(hook)
model.compile()  # this doesn't work
# model.forward = torch.compile(model.forward)  # this works

x = torch.randn(4, d_model).cuda()
model(x)
```

To check if the model is compiled, I use `TORCH_LOGS="output_code"`. For `model.compile()`, there is no code output. For `model.forward = torch.compile(model.forward)`, there is compiled triton code as expected.

Although `model.forward = torch.compile(model.forward)` is a valid workaround, it prevents me from doing full-compile a model in a function e.g. compute loss with signature `compute_loss(model, inputs, labels)`. Again, for this case, a workaround is to isolate the loss computation logic and compile it separately, but it gets messy. Ideally torch.compile should only ignore the hooks, not the `.forward()` logic, correctly.

### Error logs

_No response_

### Versions

torch==2.6.0.dev20241208+cu126

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186051

Dynamo normally graph-breaks when it calls a function wrapped with torch.compiler.disable(). For nn.Module forward pre-hooks this can happen while Dynamo is tracing nn.Module._call_impl from inside a user module's forward loop. Once that graph break occurs in the loop, Dynamo skips the parent frame, so module.compile() can miss the parent work and only capture a child module graph.

Handle the safe subset of this case by recognizing disabled forward_pre_hooks loaded from an actual module _forward_pre_hooks source whose original Python bytecode is statically pass/return None. Those hooks have no observable effect, so Dynamo can model the hook call as returning None and continue tracing the parent forward. Non-no-op disabled hooks keep the existing graph-break/eager behavior, which preserves side effects, input mutation, replacement inputs, and autograd semantics.

The fix intentionally uses the Source structure rather than matching source debug-name strings, so unrelated locals named _forward_pre_hooks still honor torch.compiler.disable().

Fixes #142358

Generated by my agent

Test Plan:

python test/dynamo/test_hooks.py HooksTests.test_disabled_noop_named_forward_pre_hooks_still_graph_breaks HooksTests.test_disabled_child_forward_pre_hooks_trace_parent_forward HooksTests.test_disabled_child_forward_pre_hook_state_mutation_graph_breaks HooksTests.test_disabled_child_forward_pre_hook_input_mutation_graph_breaks HooksTests.test_disabled_child_forward_pre_hook_returning_input_grad

python test/dynamo/test_hooks.py HooksTests.test_global_module_forward_pre_hook

python test/dynamo/test_modules.py -k test_prepend_hook_ordering

python test/dynamo/test_repros.py ReproTests.test_inbuilt_nn_module_forward_after_hook_graph_break

lintrunner -a

git diff --cached --check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
